common/state_machine.hpp

```diff
@@ -0,0 +1,231 @@
+
+#ifndef MENDER_COMMON_STATE_MACHINE_HPP
+#define MENDER_COMMON_STATE_MACHINE_HPP
+
+#include <queue>
+#include <unordered_map>
+#include <unordered_set>
+
+#include <common/events.hpp>
+#include <common/log.hpp>
+
+namespace mender {
+namespace common {
+namespace state_machine {
+
+using namespace std;
+
+namespace events = mender::common::events;
+namespace log = mender::common::log;
+
+template <typename ContextType, typename EventType>
+class StateMachineRunner;
+
+template <typename EventType>
+class EventPoster {
+public:
+	virtual ~EventPoster() {
+	}
+
+	virtual void PostEvent(EventType event) = 0;
+};
+
+template <typename ContextType, typename EventType>
+class State {
+public:
+	virtual ~State() {
+	}
+
+	virtual void OnEnter(ContextType &ctx, EventPoster<EventType> &poster) = 0;
+};
+
+enum class TransitionFlag {
+	Immediate,
+	Deferred,
+};
+
+template <typename ContextType, typename EventType>
+class StateMachine {
+public:
+	StateMachine() = default;
+	StateMachine(StateMachine &) = delete;
+
+private:
+	struct TransitionCondition {
+		// Note: Comparing address-of states. We don't want to rely on comparison operators
+		// in the states themselves, and we just want to know if they are different
+		// instances.
+		State<ContextType, EventType> *state;
+		EventType event;
+
+		bool operator==(const TransitionCondition &t) const {
+			return state == t.state && event == t.event;
+		}
+	};
+
+	class Hasher {
+	public:
+		size_t operator()(const TransitionCondition &obj) const {
+			return std::hash<State<ContextType, EventType> *>()(obj.state)
+				   ^ std::hash<int>()(static_cast<int>(obj.event));
+		}
+	};
+
+	State<ContextType, EventType> *current_state_;
+	unordered_map<TransitionCondition, State<ContextType, EventType> *, Hasher> transitions_;
+	unordered_set<EventType> deferred_events_;
+
+	friend class StateMachineRunner<ContextType, EventType>;
+
+public:
+	void AddTransition(
+		State<ContextType, EventType> &source_state,
+		EventType event,
+		State<ContextType, EventType> &target_state,
+		TransitionFlag flag) {
+		transitions_[TransitionCondition {&source_state, event}] = &target_state;
+		if (flag == TransitionFlag::Deferred) {
+			// Event is involved in at least one deferred transition, so mark that.
+			deferred_events_.insert(event);
+		}
+	}
+};
+
+template <typename ContextType, typename EventType>
+class StateMachineRunner : virtual public EventPoster<EventType> {
+public:
+	StateMachineRunner(ContextType &ctx) :
+		ctx_(ctx) {
+	}
+	StateMachineRunner(StateMachineRunner &) = delete;
+
+	void PostEvent(EventType event) override {
+		event_queue_.push(event);
+		PostToEventLoop();
+	}
+
+	void RunOne() {
+		if (event_queue_.empty()) {
+			return;
+		}
+
+		size_t size = event_queue_.size();
+		vector<State<ContextType, EventType> *> to_run;
+
+		for (size_t count = 0; count < size; count++) {
+			bool deferred = false;
+			auto event = event_queue_.front();
+			event_queue_.pop();
```

Comment:
I find it weird that there is no `pop_front()`. Hit this took when I wanted to use `queue` for the `Authenticator::pending_actions_`.

Reply:
I know. Especially since some of the other structures which are more flexible have both. So it would mirror the interfaces better if `queue` also had it.